src/mongocxx/test/spec/monitoring.cpp

```diff
@@ -267,6 +275,10 @@ void apm_checker::set_command_succeeded(options::apm& apm) {
                            make_document(kvp("reply", event.reply()),
                                          kvp("command_name", event.command_name()),
                                          kvp("operation_id", event.operation_id()))));
+
+        // We MUST NOT report a service_id() outside of load-balancing mode:
+        CHECK_FALSE(event.service_id());
```

Comment:
Nice! Can you add positive test cases that with mocks?

I am thinking like the following pseudo-code for each of the event types:
```
bool apm_command_started_get_service_id_called = false;
bson_oid_t mock_oid;

// TODO: initialize mock_oid

auto apm_command_started_get_service_id =
	libmongoc::apm_command_started_get_service_id.create_instance();
apm_command_started_get_service_id->interpose([]() {
    apm_command_started_get_service_id_called = true;
    return &mock_oid;
})

/* Register command monitoring callbacks. */
bool on_command_started_called = false;
auto apm_opts;
apm_opts.on_command_started ([](mongocxx::events::command_started_event* event) {
   on_command_started_called = true;
   // TODO: compare event.service_id () to mock_oid's representation
})


// Add necessary options for versioned API test environment.
auto client_opts;
mongocxx::test_util::add_test_server_api(client_opts);

// Add command monitoring callbacks.
client_opts.apm_opts (apm_opts);

auto client = mongocxx::client (uri() test_util::add_test_server_api()});
// Send a command (e.g. ping)
client["db"].command(/* ... */)

CHECK_TRUE (apm_command_started_get_service_id_called);
CHECK_TRUE (on_command_started_called);

Reply:
Working on this! I'm still trying to get my head around how the test harness works; some of it seems straightforward, but there are a few things I'll probably have to ask about when you have time.